File: packages/zephyr_re/protocol/packets.py
"""Hazel APK packet builders for fan and lighting (product API)."""

from __future__ import annotations

from zephyr_re.constants import EXTERNAL_LIGHT_ZONE, INTERNAL_LIGHT_ZONE
from zephyr_re.protocol.base import FanSpeed

FAN_LEVEL: dict[FanSpeed, int] = {
    FanSpeed.OFF: 0,
    FanSpeed.LOW: 1,
    FanSpeed.HIGH: 2,
}

EFFECT_TYPE: dict[str, int] = {
    "off": 0,
    "static": 1,
    "breathing": 2,
    "spectrum": 3,
    "wave": 4,
}

WAVE_RATE_FACTORY = 90
WAVE_RATE_SLOW = 15
WAVE_RATE_MEDIUM = 50
WAVE_RATE_FAST = 100

WAVE_SPEED_PRESETS: dict[str, int] = {
    "factory": WAVE_RATE_FACTORY,
    "slow": WAVE_RATE_SLOW,
    "medium": WAVE_RATE_MEDIUM,
    "fast": WAVE_RATE_FAST,
}


def effect_set_header(req_id: int, payload: bytes, *, sub_id: int = 0, zone: int) -> bytes:
    plen = len(payload)
    return bytes(
        [
            req_id & 0xFF,
            plen & 0xFF,
            (plen >> 8) & 0xFF,
            (plen >> 16) & 0xFF,
            0x10,
            0x03,
            sub_id & 0xFF,
            zone & 0xFF,
        ]
    )


def ble_effect_configuration(
    effect_type: int,
    param1: int = 0,
    param2: int = 0,
    *,
    color1: tuple[int, int, int] | None = None,
    color2: tuple[int, int, int] | None = None,
) -> bytes:
    body = [effect_type & 0xFF, param1 & 0xFF, param2 & 0xFF]
    if color1 is None and color2 is None:
        body.append(0)
    elif color1 is not None and color2 is None:
        body.append(1)
        body.extend(c & 0xFF for c in color1)
    else:
        body.append(2)
        if color1 is not None:
            body.extend(c & 0xFF for c in color1)
        if color2 is not None:
            body.extend(c & 0xFF for c in color2)
    return bytes(body)


def effect_off_v3_packets(
    req_id: int,
    zone: int,
    *,
    sub_id: int = 0,
) -> tuple[bytes, bytes]:
    payload = b"\x00"
    return effect_set_header(req_id, payload, sub_id=sub_id, zone=zone), payload


def effect_packets(
    req_id: int,
    effect: str,
    zone: int,
    *,
    sub_id: int = 0,
    r: int = 0,
    g: int = 0,
    b: int = 0,
    r2: int | None = None,
    g2: int | None = None,
    b2: int | None = None,
    wave_left_to_right: bool = True,
    wave_rate: int = WAVE_RATE_FACTORY,
) -> tuple[bytes, bytes]:
    name = effect.lower()
    effect_type = EFFECT_TYPE.get(name)
    if effect_type is None:
        raise ValueError(f"Unknown effect: {effect}")

    if effect_type == 0:
        return effect_off_v3_packets(req_id, zone, sub_id=sub_id)
    if effect_type == 1:
        payload = ble_effect_configuration(1, 0, 0, color1=(r & 0xFF, g & 0xFF, b & 0xFF))
    elif effect_type == 2:
        c1 = (r & 0xFF, g & 0xFF, b & 0xFF) if (r or g or b) else None
        c2 = None
        if r2 is not None and g2 is not None and b2 is not None:
            c2 = (r2 & 0xFF, g2 & 0xFF, b2 & 0xFF)
        payload = ble_effect_configuration(2, 0, 0, color1=c1, color2=c2)
    elif effect_type == 3:
        payload = ble_effect_configuration(3, 0, 0)
    elif effect_type == 4:
        direction = 1 if wave_left_to_right else 2
        payload = ble_effect_configuration(4, direction, wave_rate & 0xFF)
    else:
        raise ValueError(f"Unsupported effect type: {effect_type}")

    return effect_set_header(req_id, payload, sub_id=sub_id, zone=zone), payload


def fan_speed_packets(req_id: int, speed: FanSpeed) -> tuple[bytes, bytes]:
    level = FAN_LEVEL[speed]
    header = bytes([req_id & 0xFF, 0x06, 0x00, 0x00, 0x11, 0x01, 0x01, 0x00])
    payload = bytes([0x01, 0x00, level & 0xFF, 0x00, level & 0xFF, 0x00])
    return header, payload


__all__ = [
    "EFFECT_TYPE",
    "EXTERNAL_LIGHT_ZONE",
    "INTERNAL_LIGHT_ZONE",
    "WAVE_RATE_FACTORY",
    "WAVE_SPEED_PRESETS",
    "effect_packets",
    "fan_speed_packets",
]

File: public/js/protocol/packets.js
/** Hazel APK packet builders — ported from reverse-engineering reference. */

import {
  EXTERNAL_LIGHT_ZONE,
  INTERNAL_LIGHT_ZONE,
} from "./constants.js";

export const FAN_LEVEL = { off: 0, low: 1, high: 2 };

export const EFFECT_TYPE = {
  off: 0,
  static: 1,
  breathing: 2,
  spectrum: 3,
  wave: 4,
};

export const WAVE_RATE_FACTORY = 90;
/** Firmware wave byte: lower = faster rotation, higher = slower. */
export const WAVE_RATE_FAST = 15;
export const WAVE_RATE_MEDIUM = 50;
export const WAVE_RATE_SLOW = 100;

export const WAVE_SPEED_PRESETS = {
  slow: WAVE_RATE_SLOW,
  medium: WAVE_RATE_MEDIUM,
  fast: WAVE_RATE_FAST,
  factory: WAVE_RATE_FACTORY,
};

function effectSetHeader(reqId, payload, { subId = 0, zone }) {
  const plen = payload.length;
  return Uint8Array.from([
    reqId & 0xff,
    plen & 0xff,
    (plen >> 8) & 0xff,
    (plen >> 16) & 0xff,
    0x10,
    0x03,
    subId & 0xff,
    zone & 0xff,
  ]);
}

function bleEffectConfiguration(
  effectType,
  param1 = 0,
  param2 = 0,
  { color1 = null, color2 = null } = {},
) {
  const body = [effectType & 0xff, param1 & 0xff, param2 & 0xff];
  if (!color1 && !color2) {
    body.push(0);
  } else if (color1 && !color2) {
    body.push(1, ...color1.map((c) => c & 0xff));
  } else {
    body.push(2);
    if (color1) body.push(...color1.map((c) => c & 0xff));
    if (color2) body.push(...color2.map((c) => c & 0xff));
  }
  return Uint8Array.from(body);
}

export function effectOffPackets(reqId, zone, { subId = 0 } = {}) {
  const payload = Uint8Array.from([0x00]);
  return [effectSetHeader(reqId, payload, { subId, zone }), payload];
}

export function effectPackets(
  reqId,
  effect,
  zone,
  {
    subId = 0,
    r = 0,
    g = 0,
    b = 0,
    r2 = null,
    g2 = null,
    b2 = null,
    waveLeftToRight = true,
    waveRate = WAVE_RATE_FACTORY,
  } = {},
) {
  const name = effect.toLowerCase();
  const effectType = EFFECT_TYPE[name];
  if (effectType === undefined) {
    throw new Error(`Unknown effect: ${effect}`);
  }

  if (effectType === 0) {
    return effectOffPackets(reqId, zone, { subId });
  }

  let payload;
  if (effectType === 1) {
    payload = bleEffectConfiguration(1, 0, 0, {
      color1: [r & 0xff, g & 0xff, b & 0xff],
    });
  } else if (effectType === 2) {
    const c1 = r || g || b ? [r & 0xff, g & 0xff, b & 0xff] : null;
    const c2 =
      r2 !== null && g2 !== null && b2 !== null
        ? [r2 & 0xff, g2 & 0xff, b2 & 0xff]
        : null;
    payload = bleEffectConfiguration(2, 0, 0, { color1: c1, color2: c2 });
  } else if (effectType === 3) {
    payload = bleEffectConfiguration(3, 0, 0);
  } else if (effectType === 4) {
    const direction = waveLeftToRight ? 1 : 2;
    payload = bleEffectConfiguration(4, direction, waveRate & 0xff);
  } else {
    throw new Error(`Unsupported effect type: ${effectType}`);
  }

  return [effectSetHeader(reqId, payload, { subId, zone }), payload];
}

export function fanSpeedPackets(reqId, speed) {
  const level = FAN_LEVEL[speed];
  const header = Uint8Array.from([
    reqId & 0xff, 0x06, 0x00, 0x00, 0x11, 0x01, 0x01, 0x00,
  ]);
  const payload = Uint8Array.from([
    0x01, 0x00, level & 0xff, 0x00, level & 0xff, 0x00,
  ]);
  return [header, payload];
}
